enterprise/jaas-security/src/main/java/ClientLoginModule.java

```diff
@@ -0,0 +1,198 @@
+        } catch (Exception e) {
+            throw new LoginException(e.getClass().getName() + ":" + e.getMessage());
+        }
+
+        if(credentials == null) {
+            logger.log(Level.WARNING, "Credentials could not be retrieved!");
+            return false;
+        }
+        logger.log(Level.INFO, "Authenticating " + SecurityUtil.getCredentialsFullName(credentials));
+
+        if (credentials instanceof UsernamePasswordCredentials){
+            loginOk = doLoginCheck((UsernamePasswordCredentials) credentials);
+        }
+
+        return loginOk;
+    }
+
+    private boolean doLoginCheck(UsernamePasswordCredentials credentials) {
+
+        String username = credentials.getUsername();
+        String password = allowedUsersMap.get(username);
+        boolean loginCheckOk = false;
+
+        if (password != null){
+            if(password.equals(credentials.getPassword())){
+                String userGroup = userGroups.get(username);
+                if (userGroup != null){
+                    userGroupCredentials = new UserGroupCredentials(credentials.getEndpoint(),userGroup);
+                    sharedState.put(SecurityConstants.ATTRIBUTE_CREDENTIALS, credentials);
+                    loginCheckOk = true;
+                } else {
+                    logger.log(Level.WARNING, "User Group not found for user " + username);
+                    loginCheckOk = false;
+                }
+            }
+        } else {
+            logger.log(Level.WARNING, "User details not found for " + username);
+            loginCheckOk = false;
+        }
+
+        return loginCheckOk;
+
+    }
+
+    /**
+     * Commit is called when all of the modules in the chain have passed.
+     * @return
+     * @throws LoginException
+     */
+    public final boolean commit() throws LoginException {
+        logger.log(Level.FINEST, "Committing authentication of " + SecurityUtil.getCredentialsFullName(credentials));
+        final Principal principal = new ClusterPrincipal(userGroupCredentials);
+        subject.getPrincipals().add(principal);
+        sharedState.put(SecurityConstants.ATTRIBUTE_PRINCIPAL, principal);
+        return true;
+    }
+
+    /**
+     * Abort is called when one of the modules in the chain has failed.
+     * @return
+     * @throws LoginException
+     */
+    public final boolean abort() throws LoginException {
+        logger.log(Level.FINEST, "Aborting authentication of " + SecurityUtil.getCredentialsFullName(credentials));
+        clearSubject();
+        return true;
+    }
+
+    /**
+     * Graceful Logout
+     *
+     * @return
+     * @throws LoginException
+     */
+    public final boolean logout() throws LoginException {
+        logger.log(Level.FINEST, "Logging out " + SecurityUtil.getCredentialsFullName(credentials));
+        clearSubject();
+        return true;
+    }
+
+    /**
+     * Tidy up the Subject
+     */
+    private void clearSubject() {
```

Comment:
Login modules should only clean-up their stuff (i.e. avoid removing data introduced by other login modules). It's a minor thing in this case, but if users use code samples as a quick-start for their implementations this should be done right.

Reply:
I only clean up if the whole chain is aborted or if there is a logout event, so I think it's fine.